### PR TITLE
use UTF8 locale on Windows

### PR DESCRIPTION
**Pull request by rfomin**
* Remove m_io.c wrappers.

* Convert win_opendir.c to UTF8.

Because we're building with UCRT, the C standard library supports UTF8 on Windows. For the MSYS2 build, this works with the MSYS MinGW UCRT environment.

### CODE REVIEW

File: src/d_deh.c [hunk 1]
```diff
@@ -46,6 +44,13 @@
 
 #include "dsdhacked.h"
 
+#ifndef TRUE
```
Comment (fabiangreffrath) [inline src/d_deh.c:47]: I think we should rather fix the code to use `true` and `false` instead.

File: src/i_main.c [hunk 2]
```diff
@@ -112,6 +113,8 @@ int main(int argc, char **argv)
    myargc = argc;
    myargv = argv;
 
+   setlocale(LC_ALL, ".UTF8");
```
Comment (fabiangreffrath) [inline src/i_main.c:116]: Should we check for a return value? What do we expect if changing to UTF-8 does not work?

**Comment by fabiangreffrath** [comment]
What do I have to expect if I compile the code in MSYS2 in the MINGW32 or MINGW64 environments?

**Comment by rfomin** [comment]
> What do I have to expect if I compile the code in MSYS2 in the MINGW32 or MINGW64 environments?

Will work, only without non-Latin paths. You may need to remove `setlocale(LC_ALL, ".UTF8");`, I'll check that later.

**Comment by fabiangreffrath** [comment]
> Will work, only without non-Latin paths. You may need to remove `setlocale(LC_ALL, ".UTF8");`, I'll check that later.

`setlocale()` is a POSIX function, so I'd expect that to work. But, what if the OS doesn't support the environment that was requested as its second parameter? It will return `NULL` then, so we should at least check for that.

File: src/i_main.c (hunk 2 above)
Comment (rfomin) [reply src/i_main.c:116]: > What do we expect if changing to UTF-8 does not work?

C functions will work in ASCII mode witch is fine for us. UCRT can be run starting from Windows XP.

**Comment by rfomin** [comment]
> But, what if the OS doesn't support the environment that was requested as its second parameter?

I tested MSYS MinGW64 without UCRT, it works but can't find non-Latin path as expected. I think SDL2 won't work on OSes so old that they don't support UTF8, so we shouldn't do anything.

@Dasperal if you still have a working Windows 7 installation, could you please test the artifacts from this branch? I'm wondering if you will need to install the redistributable and what the error message will be.

File: src/d_deh.c (hunk 1 above)
Comment (rfomin) [reply src/d_deh.c]: Fixed in https://github.com/fabiangreffrath/woof/pull/859/commits/22e9340eb75c2b02a85ab49b1603a08fb8a2ddb3

Comment (fabiangreffrath) [reply src/d_deh.c]: Thanks!

**Comment by Dasperal** [comment]
@rfomin For artefacts from the [`artifacts` GitHub Workflow](https://github.com/fabiangreffrath/woof/actions/runs/3829279659) non-ASCII paths does not work.
```
E:\trunk\woof\Инсталл>woof -config "E:\trunk\woof\Инсталл\woof.cfg"
IWAD found: doom.wad
Ultimate DOOM version
                         The Ultimate DOOM Startup v2.03

Built on Jan  3 2023
M_LoadDefaults: Load system defaults.
 default file: E:\trunk\woof\╨Ш╨╜╤Б╤В╨░╨╗╨╗\woof.cfg
Warning: Cannot read E:\trunk\woof\╨Ш╨╜╤Б╤В╨░╨╗╨╗\woof.cfg -- using built-in defaults
```

Same for binaries compiled using `MSYS2/CLANG32` env.

I am using Windows 7 x64 with Simplix 21.11.10 updates.

**Comment by rfomin** [comment]
```
E:\trunk\woof\Инсталл>woof -config "E:\trunk\woof\Инсталл\woof.cfg"
IWAD found: doom.wad
```
Wait, but you start in a non-ASCII directory and it finds doom.wad? Also can you try running `chcp 65001` in the console? It looks like you have UCRT installed since it starts up at all.

**Comment by fabiangreffrath** [comment]
> Wait, but you start in a non-ASCII directory and it finds doom.wad?

Maybe because of this?
https://github.com/fabiangreffrath/woof/blob/a45ee95d570f6e2e7e7dd68091a941622aad76dc/src/d_iwad.c#L536-L537

**Comment by rfomin** [comment]
I'm running a VM image from https://github.com/magnetikonline/linux-microsoft-ie-virtual-machines but can't get non-Latin paths to work even after installing https://support.microsoft.com/kb/3118401 and VC redist.

So I think we should keep these wrappers. On the other hand, non-Latin paths are not that important, and there are not many Win7 users. Hmm, not sure what to do.

**Comment by Dasperal** [comment]
@rfomin
```
E:\trunk\woof\Инсталл>chcp 65001
Active code page: 65001

E:\trunk\woof\Инсталл>woof -config "E:\trunk\woof\Инсталл\woof.cfg"
IWAD found: doom.wad
Ultimate DOOM version
                         The Ultimate DOOM Startup v2.03

Built on Jan  3 2023
M_LoadDefaults: Load system defaults.
 default file: E:\trunk\woof\��������������\woof.cfg
Warning: Cannot read E:\trunk\woof\��������������\woof.cfg -- using built-in defaults
```

**Comment by rfomin** [comment]
@Dasperal I guess you have non Unicode font on cmd.exe. Fonts works for me:
image

It seems `setlocale(LC_ALL, ".UTF8");` does not work on Win7.

**Comment by fabiangreffrath** [comment]
> So I think we should keep these wrappers. On the other hand, non-Latin paths are not that important, and there are not many Win7 users. Hmm, not sure what to do.

We have just killed Win XP support. Should we really drop support for the next Windows release just now?

I think the `m_io` interface is quite clean, so I'd have no problems if we keep it until the next Windows release becomes obsolete 😉 . However, if we decide to keep it, I'd like to apply the "clean-up" parts of the current PR (e.g. removal of `TRUE`/ `FALSE` and the sorting of source files in `CMakeList.txt`).

**Comment by rfomin** [comment]
> Should we really drop support for the next Windows release just now?

We just drop non-Latin paths, it works fine otherwise. Anyway, I agree, we can keep `m_io.c` wrappers for now.

**Pull request closed without merging**